Prompt caching

tpaulshippy opened the pull request:
## What this does

<!-- Clear description of what this PR does and why -->

## Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Documentation
- [ ] Performance improvement

## Scope check

- [ ] I read the [Contributing Guide](https://github.com/crmne/ruby_llm/blob/main/CONTRIBUTING.md)
- [ ] This aligns with RubyLLM's focus on **LLM communication**
- [ ] This isn't application-specific logic that belongs in user code
- [ ] This benefits most users, not just my specific use case

## Quality check

- [ ] I ran `overcommit --install` and all hooks pass
- [ ] I tested my changes thoroughly
- [ ] I updated documentation if needed
- [ ] I didn't modify auto-generated files manually (`models.json`, `aliases.json`)

## API changes

- [ ] Breaking change
- [ ] New public methods/classes
- [ ] Changed method signatures
- [ ] No API changes

## Related issues

<!-- Link issues: "Fixes #123" or "Related to #123" -->
Resolves #13

crmne commented:
@tpaulshippy this would be great to have! Will you be willing to enable it on all providers?

I'll do a proper review at some point soon.

tpaulshippy commented:
Looking to implement this in our project and now I'm wondering if it should be an opt out rather than an opt in. If you are using unique prompts every time I guess it adds some cost to cache them but my guess is in most applications prompts will get repeated, especially system prompts.

crmne commented on docs/guides/prompt-caching.md:
I don't feel like it needs its own guide. It's a small feature that can be added in the chat guide.

crmne commented on lib/ruby_llm/chat.rb:
```diff
@@ -92,16 +95,22 @@ def each(&)
+    def cache_prompts(system: false, user: false, tools: false)
+      @cache_prompts = { system: system, user: user, tools: tools }
+      self
+    end
+
     def complete(&)
       @on[:new_message]&.call
-      response = @provider.complete(
-        messages,
+      params = CompletionParams.new(
+        messages: messages,
         tools: @tools,
         temperature: @temperature,
         model: @model.id,
-        connection: @connection,
-        &
+        cache_prompts: @cache_prompts.dup,
+        connection: @connection
       )
+      response = @provider.complete(params, &)
```
I don't feel like it's necessary to make a new `CompletionParams` class. Let's _remove_ complexity.

crmne replied:
@tpaulshippy yes I'd say so, just add a disable comment.

crmne commented on lib/ruby_llm/completion_params.rb:
I don't feel like it's necessary to make a new `CompletionParams` class. Let's _remove_ complexity.

crmne commented on lib/ruby_llm/provider.rb:
I don't feel like it's necessary to make a new `CompletionParams` class. Let's _remove_ complexity.

crmne commented on spec/fixtures/large_prompt.txt:
why do you need this? you can prompt cache a small prompt too.

tpaulshippy replied:
Really?

[This](https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching#cache-limitations) would indicate otherwise.

"Shorter prompts cannot be cached"

crmne replied:
My bad, I should have checked the docs before writing! In that case, you can still generate a (nonsensical) prompt like I do for the big message test. [massive_text = 'a' * 1_000_000](https://github.com/crmne/ruby_llm/blob/78482e5763a553956132769e23279b6dd99da8f8/spec/ruby_llm/chat_error_spec.rb#L73)

crmne reviewed (changes requested):
Thank you for this feature @tpaulshippy, however there are several improvements I'd like you to make before we merge this.

On top of the ones made in the comments, and the most important one, I'd like to have prompt caching implemented in all providers.

Plus I have not fully checked the logic in `providers/anthropic` but the patch seems a bit heavy-handed with the amount of changes needed at first glance. Where all changes necessary or could it be done in a simpler manner?

tpaulshippy commented:
> > Did you see [this](https://github.com/crmne/ruby_llm/pull/234#issuecomment-2962928284)? Is the request to populate the cached token counts on the response messages for OpenAI and Gemini?
> 
> Thank you for pointing that out, I had missed it. I think it would certainly be a nice addition to RubyLLM to have all providers have almost the same level of support of caching.

Ok we have a bit of a naming issue. Here's the property names we get from each provider:

Anthropic
`cache_creation_input_tokens`
`cache_read_input_tokens`

OpenAI
`cached_tokens`

Gemini
`cached_content_token_count`

My reading of the docs indicates that the OpenAI and Gemini values correspond pretty closely with the `cache_read_input_tokens` in Anthropic. 

What should we call these properties in the Message?

tpaulshippy commented:
@crmne Ok I addressed all the feedback. Only issue I'm running into is that I'm unable to get openai and gemini to actually report any cached tokens. I may need help from someone who knows those APIs better than me. The code to report their cached token counts is in place but I commented out the tests since I could not get them to work.

[Gemini docs on this](https://ai.google.dev/gemini-api/docs/caching?lang=python#implicit-caching)
[OpenAI docs on this](https://platform.openai.com/docs/guides/prompt-caching)

smerickson commented:
@tpaulshippy thanks so much for working on this!  I'd love to be able to switch to RubyLLM for my AI library needs and this was one of the features I was still waiting on.  Really appreciate you leading the charge.

tpaulshippy commented:
> @crmne Ok I addressed all the feedback. Only issue I'm running into is that I'm unable to get openai and gemini to actually report any cached tokens. I may need help from someone who knows those APIs better than me. The code to report their cached token counts is in place but I commented out the tests since I could not get them to work.

I realized that I wasn't actually calling the API twice, which would be necessary to hit the cache. But even after trying that I still couldn't get gemini/openai to report cache hits.

crmne reviewed (changes requested):
Paul, this PR is 90% VCR cassettes, but it's still a monster and I feel like the actual code is overengineered. Passing `cache_control` through 7 layers? No thanks.

However, how about _embracing_ the fact that there may be moments we want to augment each message?

```ruby
# Just add message-level params
chat.add_message(
  role: :user,
  content: "huge doc",
  params: { cache_control: { type: 'ephemeral' } }
)
```

That's it. Add `params` to Message.

We can then even add:

```ruby
chat.with_message_params(cache_control: { type: 'ephemeral' })
```

In order to apply it to every message. That's it.

Then adding a provider agnostic implementation of cache control becomes trivial.

tpaulshippy commented:
That could work for system and user messages. How about tools? 

I still think there will be some complexity in the anthropic provider that may be unavoidable.

tpaulshippy commented:
Another complication is that these params have to be nested under the content array for anthropic.

So it would actually need to be something like this:

```
chat.add_message(
  role: :user,
  content: "huge doc",
  params: { content: [{ cache_control: { type: 'ephemeral' } }] }
)
```

Right?

tpaulshippy commented:
I don't think adding the params to all messages with chat.with_message_params will generally work as you suggest.

Couple of reasons - 
1. Anthropic limits you to 4 cache breakpoints per request.
2. You generally want your cache markers on the last system or the last user message to get the maximum use of the cache.

tpaulshippy commented:
I do like the suggestion of adding params to messages because that could give more fine grained control over exactly where you want the breakpoints. Would also enable passing optional keys on other providers like this one from OpenAI --
![image](https://github.com/user-attachments/assets/cf00b894-1eb7-4956-8f3c-d18c3d3f62a7)

tpaulshippy commented on docs/_core_features/chat.md:
```diff
@@ -500,6 +500,30 @@ puts "Total Conversation Tokens: #{total_conversation_tokens}"
+chat = RubyLLM.chat(model: 'claude-3-5-haiku-20241022')
+
+# Enable caching for different types of content
+chat.cache_prompts(
```
The more I think about it, the more I'm thinking it is odd to have a provider specific feature enabled like this. I wish anthropic was just more like the others in this area.

tpaulshippy reviewed (commented):
Would like to explore making this a per message option and maybe something like a with_tools_params option to make this more provider agnostic.

tpaulshippy commented:
> One complication is that these params have to be nested under the content array for anthropic.
> 
> So it would actually need to be something like this:
> 
> ```
> chat.add_message(
>   role: :user,
>   content: "huge doc",
>   params: { content: [{ cache_control: { type: 'ephemeral' } }] }
> )
> ```
> 

Looks like to make this work, deep_merge will have to be enhanced to support arrays. Currently it just overrides the whole array rather than merging the hashes within. Will need something like this:
<img width="661" height="489" alt="image" src="https://github.com/user-attachments/assets/64976ab9-59f3-4950-8d2d-308e9390495c" />
What do you think @crmne ?

tpaulshippy commented:
> The other way I'm willing to accept is: RubyLLM.chat(cache: true) and chat.ask("message", cache: true)

And that would do what? Cache the last message, the last system prompt, and the tools? What if someone wants to just cache tools or just their system message? Setting the cache breakpoint on the user message is useful in multi-turn conversations, but in one shot prompt scenarios you only want to cache your system prompt since it is unlikely you will see the exact same user message again in a five minute window.

maximevaillancourt commented:
> Caching support is also a blocker on us making the jump to RubyLLM, thanks all!

No need to wait: use `with_params`!

```rb
RubyLLM
  .chat(model: "claude-sonnet-4-20250514")
  .with_params(system: [{
    type: "text",
    text: "This is my very long system prompt that will get cached.",
    cache_control: { type: "ephemeral" },
  }])
```

sosso commented:
> > Caching support is also a blocker on us making the jump to RubyLLM
> 
> No need to wait: use `with_params` in the meantime!
> 
> ```ruby
> RubyLLM
>   .chat(model: "claude-sonnet-4-20250514")
>   .with_params(system: [{
>     type: "text",
>     text: "This is my very long system prompt that will get cached.",
>     cache_control: { type: "ephemeral" },
>   }])
> ```

Hm, When trying that @maximevaillancourt , and later doing a .ask, my system prompt doesn't end up getting into openrouter. Are you using this approach successfully?

maximevaillancourt commented:
> Are you using this approach successfully?

Yes, but worth noting that I'm using `claude-sonnet-4-20250514` (the Anthropic one) directly, not an OpenRouter one, maybe that explains the difference in behaviour.

sosso commented:
Thanks for picking this back up!

Have you played around much with the 1h TTL, @tpaulshippy https://docs.claude.com/en/docs/build-with-claude/prompt-caching#1-hour-cache-duration

tpaulshippy commented:
> Have you played around much with the 1h TTL, @tpaulshippy

No I haven't. Didn't seem that useful for our scenarios. Would be a good addition to this library though.

sosso commented:
This new change is working almost perfectly for us, @crmne ! One callout: we're using the OpenRouter provider (but ultimately Anthropic models primarily within that), and while the message caching with raw blocks (using RubyLLM::Providers::Anthropic::Content in an openrouter chat) is working great, the tool with_params pattern (in the tool subclasses) is not caching definitions. I think it's because of the subclass hierarchy.

Also, Anthropic's a bit strange in that their docs have you cache the *last* tool, https://docs.claude.com/en/docs/build-with-claude/prompt-caching#prompt-caching-examples (Caching tool definitions).

tpaulshippy commented:
Even if it were opt in, a way to properly turn on caching for Anthropic in one line without having to track which will be your last tool seems like it would be nice.
